### PR TITLE
fix-body-parameter-required

### PR DESCRIPTION
**Pull request by qiaozha**
fixes https://github.com/Azure/autorest.typescript/issues/2677

### CODE REVIEW

File: packages/typespec-ts/test/integration/authApiKey.spec.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ describe("AuthApiKeyClient Rest Client", () => {
 
   it("should return 204 when the apiKey is valid", async () => {
     try {
-      const result = await client.path("/authentication/api-key/valid").get();
+      const result = await client.path("/authentication/api-key/valid").get({});
```
Comment (qiaozha) [inline packages/typespec-ts/test/integration/authApiKey.spec.ts:19]: do we still need this to make the test pass?

File: packages/typespec-ts/test/integration/unionBody.spec.ts [hunk 2]
```diff
@@ -26,7 +26,7 @@ describe("UnionBodyClient Rest Client", () => {
 
   it("should have no compile error for response body", async () => {
     try {
-      const result = await client.path("/response-union-body").get();
+      const result = await client.path("/response-union-body").get({});
```
Comment (qiaozha) [inline packages/typespec-ts/test/integration/unionBody.spec.ts:29]: is this missing to revert ?

File: packages/typespec-ts/src/transform/transformPaths.ts [hunk 3]
```diff
@@ -160,7 +161,11 @@ function hasRequiredOptions(
   dpgContext: SdkContext,
   routeParameters: HttpOperationParameters
 ) {
-  const isRequiredBodyParam = routeParameters.bodyParameter?.optional === false;
+  const isRequiredBodyParam =
+    routeParameters.body?.type &&
+    routeParameters.body?.property?.optional !== true
```
Comment (MaryGao) [inline packages/typespec-ts/src/transform/transformPaths.ts:166]: nit: maybe a helper here

File: packages/typespec-ts/src/transform/transformPaths.ts [hunk 4]
```diff
@@ -160,7 +161,11 @@ function hasRequiredOptions(
   dpgContext: SdkContext,
   routeParameters: HttpOperationParameters
 ) {
-  const isRequiredBodyParam = routeParameters.bodyParameter?.optional === false;
+  const isRequiredBodyParam =
```
Comment (MaryGao) [inline packages/typespec-ts/src/transform/transformPaths.ts:164]: just curious tcgc already adopted this logic so no change in modular side?

**Review by MaryGao — APPROVED** [review]
Approved with small comments.

File: packages/typespec-ts/src/transform/transformPaths.ts (hunk 3 above)
Comment (qiaozha) [reply packages/typespec-ts/src/transform/transformPaths.ts]: done

**Pull request merged**